ziti/cmd/ziti/util/certs.go

```diff
@@ -0,0 +1,150 @@
+package util
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"github.com/fullsailor/pkcs7"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"reflect"
+)
+
+func WriteCert(id string, cert []byte) (string, error) {
+	cfgDir, err := ConfigDir()
+	if err != nil {
+		return "", err
+	}
+	certsDir := filepath.Join(cfgDir, "certs")
+	if err = os.MkdirAll(certsDir, 0700); err != nil {
+		return "", errors.Wrapf(err, "unable to create ziti certs dir %v", certsDir)
+	}
+	certFile := filepath.Join(certsDir, id)
+	if err = ioutil.WriteFile(certFile, cert, 0600); err != nil {
+		return "", err
+	}
+	fmt.Printf("Server certificate chain written to %v\n", certFile)
+	return certFile, nil
+}
+
+func ReadCert(id string) ([]byte, string, error) {
+	cfgDir, err := ConfigDir()
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "couldn't get config dir while reading cert for %v", id)
+	}
+	certFile := filepath.Join(cfgDir, "certs", id)
+	_, err = os.Stat(certFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, "", nil
+		}
+		return nil, "", errors.Wrapf(err, "error while statting cert file for %v", id)
+	}
+	result, err := ioutil.ReadFile(certFile)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "error while reading cert for %v", id)
+	}
+	return result, certFile, nil
+}
+
+func IsServerSelfSigned(host string) (bool, error) {
+	resp, err := http.DefaultClient.Get(fmt.Sprintf("%v/.well-known/est/cacerts", host))
+	if err != nil {
+		if ue, ok := err.(*url.Error); ok && errors.As(ue.Err, &x509.UnknownAuthorityError{}) {
+			return true, nil
+		}
+		return false, err
+	}
+	_ = resp.Body.Close()
+	return false, nil
+}
+
+func GetServerCertChain(host string) ([]byte, []*x509.Certificate, error) {
```

Comment:
More aptly named "GetWellKnownCerts`. The certs contained in the well-known bundle aren't guaranteed to be any number of chains. It can be a set of individual roots,1 chain with intermediates, or nothing at all.

Reply:
fixed